### ISSUE TITLE
Proposal: code snippet testing for docs

### ISSUE THREAD

**Issue by alanbernstein** [pilosa/pilosa#1579]
Our docs are full of code snippets, which would ideally be kept up to date with the code. One way to encourage this is with a CI check that extracts the code snippets, runs them, and checks for errors and/or compares with expected output. There are many ways to approach this, and factors to consider, including:

- Multiple language support, potentially including our "request/response" blocks, pql, and toml
- Multi-block sessions, maintaining state over an example that spans multiple code blocks
- Tool development effort
- Docs maintenance effort
- Docs backward compatibility

After a brief discussion, the best option seems to be add a small amount of unobtrusive metadata to our existing markdown files, and write a custom tool to run the tests. I'll summarize the proposed process here.

## Prerequisites
Several website tasks are soft blockers for this work, in that it would be beneficial to understand the constraints they force before starting on the work for this project. One is enabling syntax highlighting, which is minor and probably insignificant. Others are larger and more relevant:

- Expanding the docs file and nav structure to allow multiple levels
- Language tabs
- General site redesign

These first two may are subject to Hugo constraints that we don't fully control. Therefore it is preferable to implement them first before starting work on the docs testing tool. Since the docs structure and language tabs work are also dependent on site redesign, we can not finish them in a reasonable time. Instead, I suggest we implement both of them as MVP prototypes without any (graphical) design effort - or at least revisit our research on them - to understand any constraints they may impose on the docs testing tool.

## Docs changes

The markdown files will only need minor updates.

Metadata about the test session for each code block will be added above, as toml within an html comment. This will not affect the rendered output at all, nor will it cause any backward compatibility issues with older docs versions. metadata will include `sessionid:<int>`, with the value starting at 0 and incrementing within each file. Unique IDs might be more practical (see prerequisites section). Any other toml syntax could be supported, but prefer minimal metadata.

Example:
~~~
<!-- sessionid:0 -->
```request
curl foo
```
```response
bar
```

<!--
sessionid:1
-->
```bash
curl localhost:10101/index/foo -XPOST
```

<!--
sessionid:1
-->
```go
response, err = client.Query(myindex.BatchQuery(
		myfield.Row(5),
		myfield.Row(10)))
```
~~~

Any `sessionid:<int>` line corresponds to the code block that follows. All code blocks that share a sessionid are run in sequential order, with state maintained between them. Ideally this would include any variables, changes made in a shell, and changes made to a Pilosa server. Any `output` code block corresponds to the preceding code block, similar to the way request/response currently work. request/response could potentially be replaced with shell/output, in case that simplifies anything.

Some code blocks currently have no language specified, while some have `request`/`response`. This may be a minor abuse of the [info string](https://github.github.com/gfm/#info-string), normally used in Hugo (and other tools) to determine the language for syntax highlighting. As long as the info string is set to the correct language whenever necessary, this doesn't seem to be a problem, practically. 

I suggest all code blocks be tagged, either with a proper language, or some other identifier. Looking through the existing untagged blocks, they could all be handled with these four identifiers:

- "bash" (19x, for testing pilosa subcommands, environment setup, etc)
- "output" (4x, for comparing results of previous block)
- "spec" (16x, for query-language.md (this could just as well be "ignore" for now))
- "ignore" (55x, code that can't be reasonably tested, e.g. installation, docker commands)

Alternatively, the "spec" and "ignore" blocks can be left blank, which will indicate the blocks should be ignored.

## Docs build process
No change for this approach. Additional steps might be necessary, if we moved docs to org files or code files. Note that Hugo does [support](https://discourse.gohugo.io/t/org-mode-content-support-merged-into-hugo-master/5530) org directly (as does github), but it may not be as mature as markdown.

## Test runner tool
The tool, written in go, will read each markdown file, naively parse html comments and markdown code fences, and create sessions and "blocks" (one per markdown code block). A session consists of multiple blocks, all of the same type. Each session may require some setup, such as a fresh Pilosa server, or a clean working directory. For a language session, variables should be maintained across code blocks (low priority).

The tool will be designed to handle new types of code blocks in a modular way, so it can be built to handle only one type (request/response or bash/output) at first, then expanded as necessary. 

As this tool is dependent on the rest of the process, I'd like to see some comments on that before developing it much further.

## Numbers

Currently, the docs have 321 total code blocks, split up like this:

| tag      |  # |
|---------|-----|
| request  | 95 |
| response | 94 |
| bash*    | 19 |
| go       |  5 |
| ignore*  | 55 |
| java     |  1 |
| json     |  1 |
| output*  |  4 |
| pql      |  3 |
| python   |  2 |
| spec*    | 16 |
| sql      |  1 |
| toml     | 27 |
| yaml     |  1 |

`*` = proposed, currently untagged

| file                |  # |
|----------------|---------|
| administration.md   | 11 |
| api-reference.md    | 23 |
| client-libraries.md |  6 |
| configuration.md    | 24 |
| console.md          |  2 |
| data-model.md       | 10 |
| examples.md         | 20 |
| getting-started.md  | 27 |
| installation.md     | 32 |
| pdk.md              |  1 |
| query-language.md   | 73 |
| tutorials.md        | 92 |

**Comment by alanbernstein** [comment]
I made an attempt at this, and unsurprisingly found that there were too many special cases to be able to support this properly. The point was to automatically check that code snippets run as expected, but the amount of effort to maintain this test runner defeats the purpose.
